### ISSUE TITLE
Interpolator: Use more advanced interpolation methods

### ISSUE THREAD

**Issue by Jammy2211** [rhayes777/PyAutoFit#686]
The interpolator now works for the CTI use case, here is a simple boiled-down example which is a simplified version of (https://github.com/Jammy2211/autocti_workspace_test/blob/main/temporal/model.py):

```
"""
Temporal: Individual Fits
=========================

In this script, we will fit multiple charge injection imaging to calibrate CTI, where:

 - The CTI model consists of one parallel `TrapInstantCapture` species.
 - The `CCD` volume filling is a simple parameterization with just a `well_fill_power` parameter.
 - The `ImagingCI` is simulated with uniform charge injection lines and no cosmic rays.

The multiple datasets are representative of CTI calibration data taken over the course of a space mission. Therefore,
the model-fitting aims to determine the increase in the density of traps with time.

This script fits each dataset one-by-one and uses the results post-analysis to determine the density evolution
parameters. Other scripts perform more detailed analyses that use more advanced statistical methods to provide a
better estimate.
"""
import numpy as np
from os import path
import autofit as af
import autocti as ac

"""
We now load every image, noise-map and pre-CTI charge injection image as instances of the `ImagingCI` object.

We load and fit each dataset, accquried at different times, one-by-one. We do this in a for loop to avoid loading 
everything into memory.
"""
instance_list = []

time_list = range(0, 10)

for time in time_list:

    dataset_time = f"time_{time}"
    dataset_time_path = path.join(dataset_path, dataset_time)

    imaging_ci_list = [
        ac.ImagingCI.from_fits(
            image_path=path.join(dataset_time_path, f"data_{int(norm)}.fits"),
            noise_map_path=path.join(
                dataset_time_path, f"norm_{int(norm)}", "noise_map.fits"
            ),
        )
        for layout, norm in zip(layout_list, norm_list)
    ]

    """
    __Model__
    
    We now compose our CTI model, which represents the trap species and CCD volume filling behaviour used to fit the 
    charge  injection data. In this example we fit a CTI model with:
    
     - One parallel `TrapInstantCapture`'s which capture electrons during clocking instantly in the parallel direction
     [2 parameters].
    
     - A simple `CCD` volume filling parametrization with fixed notch depth and capacity [1 parameter].
    
    The number of free parameters and therefore the dimensionality of non-linear parameter space is N=3.
    """
    parallel_trap_0 = af.Model(ac.TrapInstantCapture)
    parallel_traps = [parallel_trap_0]
    parallel_ccd = af.Model(ac.CCDPhase)
    parallel_ccd.well_notch_depth = 0.0
    parallel_ccd.full_well_depth = 200000.0

    model = af.Collection(
        cti=af.Model(
            ac.CTI2D,
            parallel_trap_list=[parallel_trap_0],
            parallel_ccd=parallel_ccd,
        ),
        time=time, # <- This is key to the interpolator
    )

    """
    __Search__
    
    The CTI model is fitted to the data using a `NonLinearSearch`. In this example, we use the
    nested sampling algorithm Dynesty (https://dynesty.readthedocs.io/en/latest/).
    """
    search = af.DynestyStatic(
        path_prefix=path.join(dataset_label, dataset_time),
        name="parallel[x1]",
        nlive=50,
    )

    """
    __Analysis__
    
    The `AnalysisImagingCI` object defines the `log_likelihood_function` used by the non-linear search to fit the 
    model to the `ImagingCI` dataset.
    """
    analysis_list = [
        ac.AnalysisImagingCI(dataset=imaging_ci)
        for imaging_ci in imaging_ci_list
    ]
    analysis = sum(analysis_list)
    analysis.n_cores = 1

    """
    __Model-Fit__
    
    We can now begin the model-fit by passing the model and analysis object to the search, which performs a non-linear
    search to find which models fit the data with the highest likelihood.
    """
    result_list = search.fit(model=model, analysis=analysis)

    instance_list.append(result_list[0].instance)

"""
We now use the interpolator to create a new instance of the model, where the trap densities
are at time=1.5.
"""

interpolator = af.LinearInterpolator(instances=instance_list)
instance = interpolator[interpolator.time == 1.5]

print(instance.cti.parallel_trap_list[0].density)

```

**Interpolation Method:**

For Euclid, we do not want to assume `y = mx + c` but instead need a more general interpolation tool.

This would either be some sort of spline fit, or an equation with more parameters.

We also need the interpolation to factor in the inferred errors on each model fit, which I believe we currently do not do.

@rjmassey do you have any suggestions for how to do this, or what algorithm to you?

**Comment by rjmassey** [comment]
TLDR: don't know until we get the data. But could we have a hierarchical model in which beta has a universal value; ratios of trap densities vary slowly over time (as splines?); and the total trap density is y=mt+c within specific ranges of t, but c jumps between them.

There are two big uncertainties:
1) Either all the data shows a smooth evolution for the entire mission duration, and can be modelled in a single function, or it can only be modelled in chunks. I expect Euclid to be only piecewise smooth, because of e.g. Solar events (but I expected the same for Hubble, and that's been smooth for 2 decades).

Assuming it will be only piecewise smooth, we still have a choice: we could either look at data +/-n days from the date in question - that's essentially a spline fit, as you sugest. However, we will need to choose n - and whatever we choose, that'll give a bad fit near any discontinuity. I would instead suggest a model with lots of (y=mt+c) models, each within a range t1<t<t2, where t1 and t2 are times of discontinuity events found from an initial nonlinear search, i.e. t2 for one period is the same as t1 for the next period. Is that feasible?

2) The other issue is that we expect some parameters to be constant (within measurement uncertainty). For example, the values of beta and the *relative* density of trap species were constant for Hubble. It is probably be worth treating these differently from the parameters that we expect to grow over time, such as the *sum of* trap desnities (which can be modelled as the piecewise y=mt+c or otherwise). There could be several ways of dealing with this, but my intuition is that it could be as simple as retaining the covariance between parameters, so we can include them in the fit, to get a single value of beta for each continuous period. The best approach would be to use a hierarchical model in which beta has a universal value; ratios of trap densities vay slowly over time; and the total trap density varies as y=mt+c within specific ranges of t, but c jumps between them.

**Comment by rhayes777** [comment]
I've added a CubicSpline interpolator here:

https://github.com/rhayes777/PyAutoFit/pull/687

From what RIchard said it sounds like it might make sense to instead use a sliding window around the time we're interested in and only for those attributes which we know vary over time. If the relationship tends to be linear locally then the existing LinearInterpolator may be sufficient.

**Comment by rhayes777** [comment]
Could be worth having a call about this if you have time

**Comment by rjmassey** [comment]
Yep, can talk now if you like

**Comment by rhayes777** [comment]
Could you email a zoom link?

**Comment by Jammy2211** [comment]
Going to close this for now, as we are not actively working on it, but I am sure CTI will motivate us to come back to this inthe future.
